Implementation of Focal loss

shikhargoswami opened the pull request:
Focal loss was introduced in the RetinaNet paper (https://arxiv.org/pdf/1708.02002.pdf). 

Focal loss is useful for classification when you we highly imbalanced classes. It down-weights well-classified examples and focuses on hard examples. The loss value is much high for a sample which is misclassified by the classifier as compared to the loss value corresponding to a well-classified example. 

Used in single-shot object detection where the imbalance between the background class and other classes is extremely high.

Here's it's tensorflow implementation (https://github.com/tensorflow/addons/blob/v0.12.0/tensorflow_addons/losses/focal_loss.py#L26-L81)
### PR Checklist

- [x] Tests are added
- [x] Entry in NEWS.md
- [x] Documentation, if applicable
- [ ] Final review from `@dhairyagandhi96` (for API changes).

darsnack commented on src/losses/functions.jl:
```diff
@@ -425,6 +425,23 @@ function tversky_loss(ŷ, y; β=ofeltype(ŷ, 0.7))
 end
 
+"""
+    focal_loss(yhat, y; dims=1, agg=mean, gamma=2.0, eps = eps(eltype(yhat))
```
```suggestion
    focal_loss(ŷ, y; dims=1, agg=mean, γ=2.0, eps=epseltype(yhat))
```

darsnack commented on src/losses/functions.jl:
```diff
@@ -425,6 +425,23 @@ function tversky_loss(ŷ, y; β=ofeltype(ŷ, 0.7))
+"""
+    focal_loss(yhat, y; dims=1, agg=mean, gamma=2.0, eps = eps(eltype(yhat))
+
+         𝛄: modulating factor.
```
I don't know how you entered "gamma" but it's not the right unicode character (some other variant than the standard). You can enter the correct character in Julia REPL or VS Code with `\gamma[TAB]`.

shikhargoswami replied:
I searched the unicode alternative in this Julia Unicode input doc(https://docs.julialang.org/en/v1/manual/unicode-input/).
It comes from \bfgamma[TAB]

darsnack replied:
I would just use `\gamma`.

shikhargoswami replied:
Okay. Changing now

CarloLucibello replied:
the docstring should be homogenized to that of other losses and report the definition

CarloLucibello replied:
although in this case maybe is a bit too long and can avoid it. In any case that gamma shouldn't stand there in the middle of nothing

darsnack commented on src/losses/functions.jl:
```diff
@@ -425,6 +425,23 @@ function tversky_loss(ŷ, y; β=ofeltype(ŷ, 0.7))
+
+Return the [focal_loss](https://arxiv.org/pdf/1708.02002.pdf)
+Extremely useful for classification when you have highly imbalanced classes. It down-weights
+well-classified examples and focuses on hard examples. Loss is much high for misclassified points as compared to well-classified points. Used in single-shot detectors.
```
```suggestion
well-classified examples and focuses on hard examples.
Loss is much high for misclassified points as compared to well-classified points.
Used in single-shot detectors.
```

darsnack commented on src/losses/functions.jl:
```diff
@@ -425,6 +425,23 @@ function tversky_loss(ŷ, y; β=ofeltype(ŷ, 0.7))
+"""
+function focal_loss(ŷ, y; dims=1, agg=mean, 𝛄=2.0, eps = eps(eltype(ŷ)))
+    ŷ = ŷ .+ eps
+    p_t = [y==1 ? ŷ : 1-ŷ for (ŷ, y) in zip(ŷ, y)]
```
```suggestion
    p_t = [isone(y) ? ŷ : 1 - ŷ for (ŷ, y) in zip(ŷ, y)]
```

DhairyaLGandhi replied:
The iterating variables should be different from the names of the arrays

shikhargoswami replied:
Committing all the asked changes in a bit.

CarloLucibello replied:
comprehension is not the right way to do it, makes the loss not gpu-friendly

DhairyaLGandhi replied:
Ref https://github.com/FluxML/Flux.jl/pull/1489#issuecomment-770243087

darsnack reviewed (changes requested):
Thanks for this contribution! Left some small comments, but the approach looks good to me.

DhairyaLGandhi commented:
Rather than loop iteration, it might be faster to do array operations. That might be faster for when you actually have to materialize the result with bigger arrays

CarloLucibello commented:
That paper is cited enough that it may be worth having this loss in flux, although pytorch doesn't have it and tensorflow has it only as an addon

CarloLucibello commented on src/losses/functions.jl:
```diff
@@ -425,6 +425,25 @@ function tversky_loss(ŷ, y; β=ofeltype(ŷ, 0.7))
+         𝛄: modulating factor.
+
+Return the [focal_loss](https://arxiv.org/pdf/1708.02002.pdf)
+Extremely useful for classification when you have highly imbalanced classes. It down-weights
```
```suggestion
Can be used in classification tasks in the presence of highly imbalanced classes. It down-weights
```
let's not overstate

DhairyaLGandhi replied:
Agreed

darsnack commented:
Yeah having never heard of this loss before, I checked the paper. It's well-cited, so good enough for me to include in Flux.

shikhargoswami commented:
@darsnack @CarloLucibello  Need help on converting the list compreshension to array operations. Any useful leads?

darsnack commented:
@shikhargoswami You might take a look at how `binarycrossentropy` is written to see how to address @DhairyaLGandhi's comment.

I didn't see a performance difference when testing, but I didn't use large array sizes. Even if there isn't a gap in performance, it would be good to use the same style as the other loss functions.

CarloLucibello commented:
assuming 0/1 you can write
```julia 
p_t = y .* ŷ  + (1 .- y) .* (1 .- ŷ)  
```
So this is used only for binary classification? This should be mentioned

CarloLucibello commented on src/losses/functions.jl:
```diff
@@ -425,6 +425,25 @@ function tversky_loss(ŷ, y; β=ofeltype(ŷ, 0.7))
+Return the [focal_loss](https://arxiv.org/pdf/1708.02002.pdf)
+Extremely useful for classification when you have highly imbalanced classes. It down-weights
+well-classified examples and focuses on hard examples.
+Loss is much high for misclassified points as compared to well-classified points.
```
this sentence can be deleted, states something quite trivial

CarloLucibello commented on test/losses.jl:
```diff
@@ -174,3 +174,16 @@ end
+      0 1]
+ŷ1 = [0.6 0.3
+      0.3 0.1
+      0.1 0.6]
```
all these variables should be defined inside the testset

shikhargoswami commented:
I guess I only implemented Binary classification. Here are the changes made:
- [x] List comprehension -> Array operation
- [x] Implemented categorical_focal_loss and added it's test
- [x] Minor changes in docstring

Check if it needs any other change

darsnack commented on src/losses/functions.jl:
```diff
@@ -425,7 +425,35 @@ function tversky_loss(ŷ, y; β=ofeltype(ŷ, 0.7))
+"""
+function binary_focal_loss(ŷ, y; dims=1, agg=mean, γ=2.0, eps = epseltype(ŷ))
+    ŷ = ŷ .+ eps
+    p_t = y .*ŷ  + (1 .- y) .* (1 .- ŷ)
```
```suggestion
    p_t = y .* ŷ  + (1 .- y) .* (1 .- ŷ)
```

darsnack commented on src/losses/functions.jl:
```diff
@@ -425,7 +425,35 @@ function tversky_loss(ŷ, y; β=ofeltype(ŷ, 0.7))
+
+"""
+function categorical_focal_loss(ŷ, y; dims=1, agg=mean, γ=2.0, eps = epseltype(ŷ))
+    ŷ = softmax(ŷ; dims=dims)
+    ŷ = ŷ .+ eps
+    ce = -y .* log.(ŷ)
+    weight = (1 .- ŷ) .^ γ
```
I'm not familiar with the magnitude of the numerical stability issue, but I wonder if `logsoftmax` is better to use here even if it requires more work.

Maybe someone with a better understanding can comment on the trade-off. @CarloLucibello ?

shikhargoswami replied:
@darsnack yhat is being used at two places. In computing ce (with log) and in weight (without log)

darsnack replied:
Yeah that's why I suggested there is a trade-off. It would require computing the same thing twice but potentially be more numerically stable. I don't know enough about the stability issue to know what's the right call though.

DhairyaLGandhi replied:
We can do the efficient thing by default and switch if it's unstable

CarloLucibello replied:
The numerical stability issue comes around quite often. We could try with 
```julia
ŷ = logsoftmax(ŷ; dims=dims)
loss = @. -y  * (1 - exp(ŷ))^γ * ŷ 
```
Hopefully won't lose much in performance and it's stable

DhairyaLGandhi replied:
Do you have references to issues with log? I'd rather not lose performance in the average case here.

shikhargoswami replied:
> The numerical stability issue comes around quite often. We could try with
> 
> ```julia
> ŷ = logsoftmax(ŷ; dims=dims)
> loss = @. -y  * (1 - exp(ŷ))^γ * ŷ 
> ```
> 
> Hopefully won't lose much in performance and it's stable

This would be for the logit variant of the loss but based on the suggestions of all of you, I guess we are not implementing the variant so it is anyway scraped.

darsnack commented on src/losses/Losses.jl:
```diff
@@ -18,7 +18,8 @@ export mse, mae, msle,
     hinge_loss, squared_hinge_loss,
-    ctc_loss
+    ctc_loss,
+    focal_loss
```
This needs to be updated for the two variants.

darsnack reviewed (changes requested):
Personally, I would prefer the names `focal_loss`/`binary_focal_loss` to match `crossentropy`/`binarycrossentropy`.

ToucheSir commented on src/losses/functions.jl:
```diff
@@ -425,7 +425,35 @@ function tversky_loss(ŷ, y; β=ofeltype(ŷ, 0.7))
+    γ: modulating factor.
+
+"""
+function binary_focal_loss(ŷ, y; dims=1, agg=mean, γ=2.0, eps = epseltype(ŷ))
```
`γ` also needs to be converted to `ŷ`'s eltype.

CarloLucibello commented on src/losses/functions.jl:
```diff
@@ -425,7 +425,35 @@ function tversky_loss(ŷ, y; β=ofeltype(ŷ, 0.7))
+"""
+    categorical_focal_loss(ŷ, y; dims=1, agg=mean, γ=2.0, eps = epseltype(ŷ))
+    Softmax version of Focal Loss
+    γ: modulating factor.
```
this doesn't look like a proper docstring. These 2 lines shouldn't be indented, the paper should be referenced, it should be wordier. Please take your time to observe how other docstrings are implemented in flux so that we don't have to go over and over with review comments. There is no hurry here.

shikhargoswami replied:
@CarloLucibello I guess for class imbalance purpose, only the binary classification loss is implemented in the paper. Do I put the reference of the paper?

DhairyaLGandhi replied:
Binary is a special case of Multiclass, so do we gain much from two separate definitions at all

shikhargoswami replied:
@DhairyaLGandhi I think two separate definitions are usually implemented to achieve numerical efficiency and stability as in `crossentropy/binarycrossentropy`. The `ce` in `binarycrossentropy` is a dense array where all the elements are equal in `dims=1` so just taking the `mean` would suffice. But in `crossentropy` case, `ce` is a sparse(same as one-hot) matrix where we first `sum` in `dims=1` and take `mean`.

It just made me realise that my implementation is for logit outputs in categorical case. I guess maybe I need to implement all the four cases binary/categorical/logitbinary/logitcategorical. 

Are the logit ones really necessary(again, might be useful for numerical stability)?

darsnack replied:
I think having binary and multi-class makes sense, because the paper has binary. People using the loss will refer to the paper and expect binary (i.e. not one hot) inputs.

I would not implement the logit variants and make the docstring clear that the input should be￼ unnormalized. No need to have name-explosion over a relatively rare loss function.

shikhargoswami replied:
Agreed.

darsnack commented on src/losses/functions.jl:
```diff
@@ -425,7 +425,41 @@ function tversky_loss(ŷ, y; β=ofeltype(ŷ, 0.7))
+Return the [focal_loss](https://arxiv.org/pdf/1708.02002.pdf)
+For γ=0, the loss is mathematically equivalent to crossentropy(ŷ, y)
+
+See also: [`binary_focal_loss`](@ref)
```
```suggestion
See also: [`Losses.binary_focal_loss`](@ref) for binary (not one-hot) labels
```

shikhargoswami replied:
@darsnack Done. Can I commit the final changes now?

darsnack reviewed (changes requested):
Some suggestions on the docstrings.

darsnack commented:
I would commit the changes that address the comments above. I think this is close to being ready for approval, but we'll need to come to a consensus on the numerical stability vs. performance issue. So probably we won't be able to merge today.

DhairyaLGandhi commented:
I'll also need to review the api before we make a final call

darsnack commented on src/losses/functions.jl:
```diff
@@ -425,7 +425,47 @@ function tversky_loss(ŷ, y; β=ofeltype(ŷ, 0.7))
+See also: [`Losses.focal_loss`](@ref) for multi-class setting
+
+"""
+function binary_focal_loss(ŷ, y; dims=1, agg=mean, γ=ofeltype(ŷ, 2.0), ϵ=epseltype(ŷ))
```
I don't think this should have the `dims` keyword if it isn't used anywhere.

darsnack commented on src/losses/functions.jl:
```diff
@@ -425,7 +425,47 @@ function tversky_loss(ŷ, y; β=ofeltype(ŷ, 0.7))
+See also: [`Losses.binary_focal_loss`](@ref) for binary (not one-hot) labels
+
+"""
+function focal_loss(ŷ, y; dims=1, agg=mean, γ=ofeltype(ŷ, 2.0), ϵ=epseltype(ŷ))
```
`binary_focal_loss` implements the softmax for the user (accepting unnormalized input), but `focal_loss` doesn't. I think you want to add a `softmax` operation here.

shikhargoswami replied:
Both the functions don't implement softmax, they take normalized ŷs(which has to be the output of the last softmax layer of the model). It is in the logit loss function that we need to normalize(the user may not have used softmax layer).

darsnack replied:
You're right! No clue what I was looking at in reference to my comment.

darsnack commented on src/losses/functions.jl:
```diff
@@ -425,7 +425,47 @@ function tversky_loss(ŷ, y; β=ofeltype(ŷ, 0.7))
+"""
+function focal_loss(ŷ, y; dims=1, agg=mean, γ=ofeltype(ŷ, 2.0), ϵ=epseltype(ŷ))
+    ŷ = ŷ .+ ϵ
+    agg(sum(@. -y * (1 - ŷ)^γ * log(ŷ); dims=1))
```
In addition to the comments about softmax.
```suggestion
    agg(sum(@. -y * (1 - ŷ)^γ * log(ŷ); dims=dims))
```

darsnack commented on test/losses.jl:
```diff
@@ -174,3 +174,37 @@ end
   end
 end
+
+@testset "binary_focal_loss" begin
+    y = [0  1  0
+         1  0  1]
+    ŷ = [0.268941  0.5  0.268941
+            0.731059  0.5  0.731059]
+
+    y1 = [1 0
+          0 1]
+    ŷ1 = [0.6 0.3
+          0.4 0.7]
+    @test Flux.binary_focal_loss(ŷ, y) ≈ 0.0728675615927385
+    @test Flux.binary_focal_loss(ŷ1, y1) ≈ 0.05691642237852222
+    @test Flux.binary_focal_loss(ŷ, y; γ=0.0) == Flux.binarycrossentropy(ŷ, y)
+end
+
+@testset "focal_loss" begin
+    y = [1  0  0  0  1
+         0  1  0  1  0
+         0  0  1  0  0]
+
+    ŷ = [0.0900306  0.0900306  0.0900306  0.0900306  0.0900306
+        0.244728   0.244728   0.244728   0.244728   0.244728
+        0.665241   0.665241   0.665241   0.665241   0.665241]
+    y1 = [1 0
+          0 0
+          0 1]
+    ŷ1 = [0.4 0.2
+          0.5 0.5
+          0.1 0.3]
+    @test Flux.focal_loss(ŷ, y) ≈ 1.1277571935622628
+    @test Flux.focal_loss(ŷ1, y1) ≈ 0.45990566879720157
+    @test Flux.focal_loss(ŷ, y; γ=0.0) == Flux.crossentropy(ŷ, y)
+end
```
These test cases all use normalized versions of `ŷ` which remains unmodified post-softmax. Instead, we should also have test cases where the sum along the rows isn't 1.

shikhargoswami replied:
Ref [#1489(comment)](https://github.com/FluxML/Flux.jl/pull/1489#discussion_r568355858)

darsnack commented on src/losses/functions.jl:
```diff
@@ -425,7 +425,47 @@ function tversky_loss(ŷ, y; β=ofeltype(ŷ, 0.7))
+Return the [binary_focal_loss](https://arxiv.org/pdf/1708.02002.pdf)
+which can be used in classification tasks with highly imbalanced classes.
+It down-weights well-classified examples and focuses on hard examples.
+The input, 'ŷ', is expected to be unnormalized.
```
```suggestion
The input, 'ŷ', is expected to be normalized (i.e. [`softmax`](@ref) output).
```

darsnack commented on src/losses/functions.jl:
```diff
@@ -425,7 +425,47 @@ function tversky_loss(ŷ, y; β=ofeltype(ŷ, 0.7))
+Return the [focal_loss](https://arxiv.org/pdf/1708.02002.pdf)
+which can be used in classification tasks with highly imbalanced classes.
+It down-weights well-classified examples and focuses on hard examples.
+The input, `ŷ`, is expected to be unnormalized.
```
```suggestion
The input, 'ŷ', is expected to be normalized (i.e. [`softmax`](@ref) output).
```

darsnack commented on src/losses/functions.jl:
```diff
@@ -425,7 +425,47 @@ function tversky_loss(ŷ, y; β=ofeltype(ŷ, 0.7))
 end
 
+"""
+    binary_focal_loss(ŷ, y; dims=1, agg=mean, γ=ofeltype(ŷ, 2.0), ϵ=epseltype(ŷ))
```
Need to remove `dims` from the docstring too

shikhargoswami replied:
oops sorry! Doing it now

shikhargoswami replied:
Everybody pardon me for doing silly mistakes!

darsnack reviewed (commented):
This looks good to go for me, but @DhairyaLGandhi will need to approve the final API.

DhairyaLGandhi commented on src/losses/functions.jl:
```diff
@@ -425,7 +425,47 @@ function tversky_loss(ŷ, y; β=ofeltype(ŷ, 0.7))
+See also: [`Losses.focal_loss`](@ref) for multi-class setting
+
+"""
+function binary_focal_loss(ŷ, y; agg=mean, γ=ofeltype(ŷ, 2.0), ϵ=epseltype(ŷ))
```
Is this the standard gamma or is it just the fonts tripping up because we don't want to make unicode too hard for users. Please also refer to current master about style guides around whitespace

shikhargoswami replied:
Yes. It is given in here: [Link](https://docs.julialang.org/en/v1/manual/unicode-input/). It comes from `\gamma[TAB]`

DhairyaLGandhi commented on test/losses.jl:
```diff
@@ -174,3 +174,37 @@ end
+          0.1 0.3]
+    @test Flux.focal_loss(ŷ, y) ≈ 1.1277571935622628
+    @test Flux.focal_loss(ŷ1, y1) ≈ 0.45990566879720157
+    @test Flux.focal_loss(ŷ, y; γ=0.0) ≈ Flux.crossentropy(ŷ, y)
```
Should also add gradient tests and GPU tests

shikhargoswami replied:
Any leads on how to write tests for these?

DhairyaLGandhi replied:
https://github.com/FluxML/Flux.jl/pull/1489#discussion_r569580943

DhairyaLGandhi commented on test/losses.jl:
```diff
@@ -174,3 +174,37 @@ end
+
+    ŷ = [0.0900306  0.0900306  0.0900306  0.0900306  0.0900306
+        0.244728   0.244728   0.244728   0.244728   0.244728
+        0.665241   0.665241   0.665241   0.665241   0.665241]
```
Where did you find these values?

shikhargoswami replied:
`ŷ = softmax(reshape(-7:7, 3, 5) .* 1f0)`

DhairyaLGandhi replied:
We should have a grads test in the test/cuda directory

DhairyaLGandhi replied:
Best to use that instead

DhairyaLGandhi reviewed (commented):
Thanks, I've added a few last thoughts, but this is looking good. We might want to review how we write our losses but that would be a more general shift, so keeping it consistent with the current seems sensible to me. Thanks again for the good work @shikhargoswami

CarloLucibello commented:
I think it would be better to assume logits as inputs instead of probabilities. In model-zoo we use `logitcrossentropy` everyqhere instead of crossentropy for numerical stability

darsnack commented:
This goes back to the numerical stability question. We would need to rework the implementation to make the logits version stable. If we do switch to logits, maybe the naming should be `logitfocalloss` and `logitbinaryfocalloss`. It's tough on the eyes but consistent with the cross entropy convention.

CarloLucibello commented:
the fact that we have both `crossentropy` and `logitcrossentropy` is quite unfortunate, when I revisited the loss functions I thought about having a single definition `function crossentropy(yhat, y; logits=true)` but couldn't figure out a nice deprecation path. Working with unnormalized log-probabilities is what one should want in most cases for numerical stability. Should we experiment with this keyword arg approach here?  Or we can just leave this as it is, not worth too much overthinking

darsnack commented:
I like the possibility of making "logit" a kwarg, but I think it is better in a separate PR. I think for here, let's just decide which version (logit or not) we want, and we can keep it consistent with what we already have.

DhairyaLGandhi commented:
I think the less magical the function, the better, and this makes certain contracts very implicit, like which function softmax we use. It goes against the self documenting code we expect from Flux. I'd need to see some very compelling reasons to do opaque looking keywords in Flux.

It's one reason I feel like we should remove the `agg`, it doesn't justify being in every loss function when it clearly doesn't add much to most of them.

CarloLucibello commented:
let's keep it as it is then, it's already done and `logitbinaryfocalloss` is quite horrifying :smile:

DhairyaLGandhi commented:
I agree the names... Need work

CarloLucibello commented on src/losses/functions.jl:
```diff
@@ -425,7 +425,47 @@ function tversky_loss(ŷ, y; β=ofeltype(ŷ, 0.7))
 end
 
+"""
+    binary_focal_loss(ŷ, y; agg=mean, γ=ofeltype(ŷ, 2.0), ϵ=epseltype(ŷ))
```
```suggestion
    binary_focal_loss(ŷ, y; agg=mean, γ=2, ϵ=eps(ŷ))
```
Even if not the exact signature, is informative enough

shikhargoswami replied:
Correct me if I'm wrong(and this is also a doubt). `γ=2` will get promoted to floating point when multiplied later. Is it more numerical efficient to declare `γ=2` than `γ=ofeltype(ŷ, 2.0)`?

shikhargoswami replied:
Does promotion bear an overhead cost?

darsnack replied:
No I think @CarloLucibello wants you to leave out the `ofeltype` just in the docstring. The code can remain the same.

shikhargoswami replied:
Okay! My bad :)

CarloLucibello replied:
Yes I meant just the docstring, but now that think about it `γ=2` is actually a better default since integer exponentiation takes a different and faster path

CarloLucibello commented on src/losses/functions.jl:
```diff
@@ -425,7 +425,47 @@ function tversky_loss(ŷ, y; β=ofeltype(ŷ, 0.7))
+end
 
+"""
+    focal_loss(ŷ, y; dims=1, agg=mean, γ=ofeltype(ŷ, 2.0), ϵ=epseltype(ŷ))
```
```suggestion
    focal_loss(ŷ, y; dims=1, agg=mean, γ=2, ϵ=eps(ŷ))
```
Even if not the exact signature, is informative enough

darsnack commented on test/cuda/losses.jl:
```diff
@@ -14,13 +14,24 @@ y = [1, 1, 0.]
 
   for loss in ALL_LOSSES
-    gpu_gradtest(loss, x, y)
+    gpu_autodiff_test(loss, x, y)
```
Why did you change this line? It is causing errors for the GPU tests.

shikhargoswami replied:
I didn't change. I pulled the recent changes from repo. I might have been changed idk.

shikhargoswami replied:
I didn't change it. I pulled the recent changes from repo. It might have been changed idk.

shikhargoswami replied:
I didn't change it. I pulled the recent changes from repo. It might have been changed idk.

darsnack replied:
Ah gotcha. @CarloLucibello do you know where this change is from?

darsnack replied:
Okay I think you just don't have all the changes from the repo. How did you pull these changes? Can you rebase?

CarloLucibello commented on src/losses/functions.jl:
```diff
@@ -425,7 +425,83 @@ function tversky_loss(ŷ, y; β=ofeltype(ŷ, 0.7))
+See also: [`Losses.focal_loss`](@ref) for multi-class setting
+
+"""
+function binary_focal_loss(ŷ, y; agg=mean, γ=ofeltype(ŷ, 2.0), ϵ=epseltype(ŷ))
```
```suggestion
function binary_focal_loss(ŷ, y; agg=mean, γ=2, ϵ=epseltype(ŷ))
```

CarloLucibello commented on src/losses/functions.jl:
```diff
@@ -425,7 +425,83 @@ function tversky_loss(ŷ, y; β=ofeltype(ŷ, 0.7))
+See also: [`Losses.binary_focal_loss`](@ref) for binary (not one-hot) labels
+
+"""
+function focal_loss(ŷ, y; dims=1, agg=mean, γ=ofeltype(ŷ, 2.0), ϵ=epseltype(ŷ))
```
```suggestion
function focal_loss(ŷ, y; dims=1, agg=mean, γ=2, ϵ=epseltype(ŷ))
```

shikhargoswami commented:
I think I got issues with the runtime of my os. Can i close this PR and open a new one?

darsnack commented:
Sure, but preferably don't delete your fork so that we don't lose all the work that's been put in. Or you can give @CarloLucibello or me write access to the fork. Seems like a simple rebase should fix things.

darsnack commented:
@shikhargoswami that should fix it. Can you address the remaining comments?

darsnack reviewed (approved):
Let's wait for CI to pass then I think this can merge.

shikhargoswami commented:
@darsnack Thanks a lot for the help!

darsnack commented:
No problem! Thank you for your contribution!

darsnack commented:
bors r+

darsnack commented:
bors cancel

darsnack commented:
bors r-

darsnack reviewed (changes requested):
@shikhargoswami sorry to delay, but I realized you didn't add the focal loss docstring to the actual docs in `docs/models/losses.md`.

darsnack reviewed (approved):
Thanks!

darsnack commented:
bors r+

DhairyaLGandhi commented:
Kyle, for the future, this is an API related change and the process does require a final approval from me.

darsnack commented:
Sorry, my mistake, I thought you had approved of the API. Next time I will leave the final merge to you on API changes.
